File: scripts/algorithms/I/Integer Break/Integer Break.js
// Runtime: 62 ms (Top 95.56%) | Memory: 42 MB (Top 66.94%)
var integerBreak = function(n) {

       if(n == 2) return 1;
       if(n == 3) return 2;
       if(n == 4) return 4;

       let c = ~~(n/3);
       let d = n % 3;

       if(d === 0){
           return 3 ** c;
       }

       if(d === 1){
           return 3 ** (c-1) * 4;
       }

       if(d === 2){
           return 3 ** (c) * 2;
       }

};

File: scripts/algorithms/I/Integer Break/Integer Break.java
// Runtime: 0 ms (Top 100.0%) | Memory: 40.80 MB (Top 7.7%)

class Solution {
    public int integerBreak(int n) {
        if (n <= 1) {
            return 0;
        }
        int[] memo = new int[n + 1];
        return maxProduct(n, memo);
    }
    
    private int maxProduct(int n, int[] memo) {
        if (n == 2) {
            return 1;
        }
        if (memo[n] != 0) {
            return memo[n];
        }
        int max = 0;
        for (int i = 1; i < n; i++) {
            max = Math.max(max, Math.max(i * (n - i), i * maxProduct(n - i, memo)));
        }
        memo[n] = max;
        return max;
    }
}
